### PR TITLE
Remove frame lag and jitter from near grabbed physical objects.

### PR DESCRIPTION
**Pull request by hyperlogic**
In the game loop, physics occurs before avatar update.  Before this PR, when the avatar is moved during avatar update, near grabbed objects will not pick up this move until one frame later, when the physics is run on the next update.

After this PR, near grabbed objects are adjusted to reflect any position or rotation change that occurred during the avatar update.  This removes lag and jitter in the motion of held objects during driving and re-centering.

### CODE REVIEW

**Comment by hyperlogic** [comment]
https://highfidelity.fogbugz.com/f/cases/448/Objects-held-in-hand-jitter-while-turning-walking-flying

**Comment by hyperlogic** [comment]
Test plan:
   * Enter the dev-demo domain while wearing a HMD with hand controllers
   * Ensure that Settings > Advanced controller motion is enabled.
   * grab a block in one of your hands.
   * With the right controller rotate your avatar in place.  The object in your hand should be solid relative to the hand controller blue ball.  Even if Avatar > Settings > Snap turns are enabled or disabled.
   * With the left controller strafe sideways.  The object in your hand should be solid relative to the blue ball.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 327fcc970b85e7acecc88527fb6d395269a9ef16 available at: https://gist.github.com/8cf52c528fa2ff631fda379784938a93<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8717-327fcc9.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8717-327fcc9.dmg

File: interface/src/avatar/AvatarActionHold.cpp [hunk 1]
```diff
@@ -460,3 +471,40 @@ void AvatarActionHold::deserialize(QByteArray serializedArguments) {
...
+    btTransform worldTrans = rigidBody->getWorldTransform();
+    AnimPose worldBodyPose(glm::vec3(1), bulletToGLM(worldTrans.getRotation()), bulletToGLM(worldTrans.getOrigin()));
+
+    // transform the body transform into sensor space with the prePhysics sensor-to-world matrix.
+    // then transform it back into world uisng the postAvatarUpdate sensor-to-world matrix.
+    AnimPose newWorldBodyPose = postAvatarUpdateRoomPose * prePhysicsRoomPose.inverse() * worldBodyPose;
+
+    worldTrans.setOrigin(glmToBullet(newWorldBodyPose.trans));
+    worldTrans.setRotation(glmToBullet(newWorldBodyPose.rot));
+    rigidBody->setWorldTransform(worldTrans);
```
Comment (sethalves) [inline interface/src/avatar/AvatarActionHold.cpp]: I doubt this setWorldTransform is needed.

Comment (hyperlogic) [inline interface/src/avatar/AvatarActionHold.cpp]: Turns out it is, I'm not 100% certain why.  But without this I do get occasional glitches.  I think it has to do with frames where there are no physics timesteps, which can happen somewhat rarely.

**Comment by Nex-Pro** [comment]
tested on win 10

Tested with Thoys.

### PASS

Block was as steady as I could hold it.  No glitches observed. Tried strafing spinning waving and flying.

File: interface/src/avatar/MyAvatar.cpp [hunk 2]
```diff
@@ -2257,3 +2263,25 @@ glm::vec3 MyAvatar::getAbsoluteJointTranslationInObjectFrame(int index) const {
...
+    std::lock_guard<std::mutex> guard(_holdActionsMutex);
+    auto iter = std::find(std::begin(_holdActions), std::end(_holdActions), holdAction);
+    if (iter != std::end(_holdActions)) {
+        _holdActions.erase(iter);
+    }
+}
+
+void MyAvatar::updateHoldActions(const AnimPose& prePhysicsPose, const AnimPose& postUpdatePose) {
+    std::lock_guard<std::mutex> guard(_holdActionsMutex);
+    for (auto& holdAction : _holdActions) {
```
Comment (sethalves) [inline interface/src/avatar/MyAvatar.cpp]: without a lock on the entity tree, I'm not sure this is safe.

Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: Ok, I added an entity tree write lock around these calls.

**Comment by hifi-gustavo** [comment]
Build console output for commit: dcd425a165802880321ea23a439c1854495c1159 available at: https://gist.github.com/8cf52c528fa2ff631fda379784938a93<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8717-dcd425a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8717-dcd425a.dmg

**Pull request merged**